File: ChatGPT(예진)/Cpp/통신에서 사용하는 알고리즘 코드 구현.cpp
#include <iostream>
#include <string>
#include <cmath>
#include <cstdlib>
#include <ctime>

// 해밍 코드 구현
std::string generateHammingCode(const std::string& data) {
    int n = data.length();
    int m = 0;
    while (pow(2, m) < n + m + 1) {
        m++;
    }
    std::string hammingCode(n + m, '0');
    int j = 0;
    for (int i = 0; i < n + m; i++) {
        if (i + 1 == pow(2, j)) {
            j++;
        }
        else {
            if (j < n) {  // j가 데이터의 길이를 벗어나지 않도록 확인
                hammingCode[i] = data[j];
                j++;
            }
        }
    }
    for (int i = 0; i < m; i++) {
        int p = pow(2, i) - 1;
        int count = 0;
        for (int j = p; j < n + m; j += 2 * (p + 1)) {
            for (int k = j; k < std::min(j + p + 1, n + m); k++) {
                if (hammingCode[k] == '1') {
                    count++;
                }
            }
        }
        if (count % 2 != 0) {
            hammingCode[p] = '1';
        }
    }
    return hammingCode;
}

// CRC 코드 구현
std::string generateCrcCode(const std::string& data, const std::string& divisor) {
    std::string crc = data;
    int dataLength = data.length();
    int divisorLength = divisor.length();
    for (int i = 0; i < dataLength - divisorLength + 1; i++) {
        if (crc[i] == '1') {
            for (int j = 0; j < divisorLength; j++) {
                crc[i + j] = ((crc[i + j] - '0') ^ (divisor[j] - '0')) + '0';
            }
        }
    }
    return crc;
}

// RSA 암호화 알고리즘 구현
int gcd(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

int modInverse(int a, int m) {
    int m0 = m, x0 = 0, x1 = 1;
    while (a > 1) {
        int q = a / m;
        int temp = m;
        m = a % m;
        a = temp;
        temp = x0;
        x0 = x1 - q * x0;
        x1 = temp;
    }
    return x1 < 0 ? x1 + m0 : x1;
}

std::pair<int, int> generateRsaKeyPair(int p, int q) {
    int n = p * q;
    int phi = (p - 1) * (q - 1);
    std::srand(std::time(0));
    int e = std::rand() % (phi - 2) + 2;
    while (gcd(e, phi) != 1) {
        e = std::rand() % (phi - 2) + 2;
    }
    int d = modInverse(e, phi);
    return std::make_pair(e, d);
}

int main() {
    // 예제 데이터 생성
    std::string data = "101010";
    std::string divisor = "1101";
    int p = 61;
    int q = 53;

    // 해밍 코드 결과 출력
    std::cout << "Hamming Code:" << std::endl;
    std::string hammingCode = generateHammingCode(data);
    std::cout << "Original Data: " << data << std::endl;
    std::cout << "Hamming Code: " << hammingCode << std::endl;

    // CRC 코드 결과 출력
    std::cout << "\nCRC Code:" << std::endl;
    std::string crcCode = generateCrcCode(data + std::string(divisor.length() - 1, '0'), divisor);
    std::cout << "Original Data: " << data << std::endl;
    std::cout << "CRC Code: " << crcCode << std::endl;

    // RSA 암호화 알고리즘 결과 출력
    std::cout << "\nRSA Encryption:" << std::endl;
    auto keyPair = generateRsaKeyPair(p, q);
    int plaintext = 42;
    int encryptedText = static_cast<int>(pow(plaintext, keyPair.first)) % (p * q);
    std::cout << "Plaintext: " << plaintext << std::endl;
    std::cout << "Encrypted Text: " << encryptedText << std::endl;

    // RSA 복호화 알고리즘 결과 출력
    std::cout << "\nRSA Decryption:" << std::endl;
    int decryptedText = static_cast<int>(pow(encryptedText, keyPair.second)) % (p * q);
    std::cout << "Decrypted Text: " << decryptedText << std::endl;

    return 0;
}

File: ChatGPT(예진)/Java/통신에서사용하는알고리즘코드구현.java
import java.util.Random;

public class 통신에서사용하는알고리즘코드구현 {
    // 해밍 코드 구현
    public static String generateHammingCode(String data) {
        int n = data.length();
        int m = 0;
        while (Math.pow(2, m) < n + m + 1) {
            m++;
        }
        char[] hammingCode = new char[n + m];
        int j = 0;
        for (int i = 0; i < n + m; i++) {
            if (i + 1 == Math.pow(2, j)) {
                hammingCode[i] = '0';
                j++;
            } else {
                if (j < n) {  // j가 데이터의 길이를 벗어나지 않도록 확인
                    hammingCode[i] = data.charAt(j);
                    j++;
                } else {
                    hammingCode[i] = '0';  // 데이터 길이를 벗어나면 0으로 채움
                }
            }
        }
        for (int k = 0; k < m; k++) { // 변경된 부분
            int p = (int) Math.pow(2, k) - 1;
            int count = 0;
            for (int l = p; l < n + m; l += 2 * (p + 1)) { // 변경된 부분
                for (int o = l; o < Math.min(l + p + 1, n + m); o++) { // 변경된 부분
                    if (hammingCode[o] == '1') {
                        count++;
                    }
                }
            }
            if (count % 2 == 0) {
                hammingCode[p] = '0';
            } else {
                hammingCode[p] = '1';
            }
        }
        return new String(hammingCode);
    }

    // CRC 코드 구현
    public static String generateCrcCode(String data, String divisor) {
        char[] crc = data.toCharArray();
        int divisorLength = divisor.length();
        for (int i = 0; i < data.length() - divisorLength + 1; i++) {
            if (crc[i] == '1') {
                for (int j = 0; j < divisorLength; j++) {
                    crc[i + j] = (char) (((crc[i + j] - '0') ^ (divisor.charAt(j) - '0')) + '0');
                }
            }
        }
        return new String(crc);
    }

    // RSA 암호화 알고리즘 구현
    public static int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    public static int modInverse(int a, int m) {
        int m0 = m, x0 = 0, x1 = 1;
        while (a > 1) {
            int q = a / m;
            int temp = m;
            m = a % m;
            a = temp;
            temp = x0;
            x0 = x1 - q * x0;
            x1 = temp;
        }
        return x1 < 0 ? x1 + m0 : x1;
    }

    public static int[] generateRsaKeyPair(int p, int q) {
        int n = p * q;
        int phi = (p - 1) * (q - 1);
        Random random = new Random();
        int e = random.nextInt(phi - 2) + 2;
        while (gcd(e, phi) != 1) {
            e = random.nextInt(phi - 2) + 2;
        }
        int d = modInverse(e, phi);
        return new int[]{e, n, d};
    }

    public static void main(String[] args) {
        // 예제 데이터 생성
        String data = "101010";
        String divisor = "1101";
        int p = 61;
        int q = 53;

        // 해밍 코드 결과 출력
        System.out.println("Hamming Code:");
        String hammingCode = generateHammingCode(data);
        System.out.println("Original Data: " + data);
        System.out.println("Hamming Code: " + hammingCode);

        // CRC 코드 결과 출력
        System.out.println("\nCRC Code:");
        String crcCode = generateCrcCode(data + "0".repeat(divisor.length() - 1), divisor);
        System.out.println("Original Data: " + data);
        System.out.println("CRC Code: " + crcCode);

        // RSA 암호화 알고리즘 결과 출력
        System.out.println("\nRSA Encryption:");
        int[] keyPair = generateRsaKeyPair(p, q);
        int plaintext = 42;
        int encryptedText = (int) Math.pow(plaintext, keyPair[0]) % keyPair[1];
        System.out.println("Plaintext: " + plaintext);
        System.out.println("Encrypted Text: " + encryptedText);

        // RSA 복호화 알고리즘 결과 출력
        System.out.println("\nRSA Decryption:");
        int decryptedText = (int) Math.pow(encryptedText, keyPair[2]) % keyPair[1];
        System.out.println("Decrypted Text: " + decryptedText);
    }
}
